buildserver/environment.go

```diff
@@ -0,0 +1,431 @@
+//
+// This is best-effort. If any errors occur or we do not detect a custom
+// gopath, an empty result is returned.
+func detectSourcegraphGOPATH(ctx context.Context, fs ctxvfs.FileSystem) (gopaths []string) {
+	cfg := readSourcegraphConfig(ctx, fs)
+	for _, p := range cfg.Go.GOPATH {
+		if !strings.HasPrefix(p, "/") {
+			// Assume all paths are relative to repo root.
+			p = "/" + p
+		}
+		gopaths = append(gopaths, p)
+	}
+	return
+}
+
+// sourcegraphConfig is a struct representing the Go portion of the
+// .sourcegraph/config.json file that may be placed in the repo root.
+type sourcegraphConfig struct {
+	// Go is the go portion of the configuration. Any language server can
+	// define their own schema for their language, hence the namespacing here.
+	Go struct {
+		// GOPATH is a list of GOPATHs to use for the repository. It is assumed
+		// each GOPATH string value is a path relative to the repository root.
+		//
+		// This is to support monorepos such as the ones described in
+		// https://blog.gopheracademy.com/advent-2015/go-in-a-monorepo/
+		//
+		// See https://docs.sourcegraph.com/extensions/language_servers/go#custom-gopaths-go-monorepos.
+		GOPATH []string
+
+		// RootImportPath defines what Go import path corresponds to the
+		// repository root. Effectively, the Go language server will clone the
+		// repository into $GOPATH/src/$ROOT_IMPORT_PATH when set.
+		//
+		// This overrides the heuristic-based approach to locate this
+		// information (from glide.yml or canonical import path comments) and
+		// gives you an opportunity to specify it directly.
+		//
+		// See https://docs.sourcegraph.com/extensions/language_servers/go#vanity-import-paths.
+		RootImportPath string
+	} `json:"go"`
+}
+
+// readSourcegraphConfig reads the .sourcegraph/config.json file from the
+// repository root if it exists, otherwise an empty struct value (not nil).
+func readSourcegraphConfig(ctx context.Context, fs ctxvfs.FileSystem) *sourcegraphConfig {
+	config := sourcegraphConfig{}
+	b, err := ctxvfs.ReadFile(ctx, fs, "/.sourcegraph/config.json")
+	if err != nil {
+		return &config
+	}
+	_ = json.Unmarshal(b, &config)
+	return &config
+}
+
+// determineRootImportPath determines the root import path for the Go
+// workspace. It looks at canonical import path comments and the
+// repo's original clone URL to infer it.
+//
+// It's intended to handle cases like
+// github.com/kubernetes/kubernetes, which has doc.go files that
+// indicate its root import path is k8s.io/kubernetes.
+func determineRootImportPath(ctx context.Context, originalRootURI lsp.DocumentURI, fs ctxvfs.FileSystem) (rootImportPath string, err error) {
+	if originalRootURI == "" {
+		return "", errors.New("unable to determine Go workspace root import path without due to empty root path")
+	}
+	u, err := gituri.Parse(string(originalRootURI))
+	if err != nil {
+		return "", err
+	}
+
+	if path.Join(u.Host, u.Path) == "github.com/golang/go" {
+		return "github.com/golang/go", nil
+	}
```

Comment:
why did you add this special case?

Reply:
This is a fix for a failing test. I started a code discussion about it a while back, but the link is broken. Here's the conversation:

> @-chrismwendt Why is a canonical import path with the same or fewer `parts` than `popComponents` invalid? I see a test failure when running:
> 
> ```
> $ go test github.com/sourcegraph/sourcegraph/enterprise/cmd/xlang-go/internal/server -run 'TestIntegration'
> --- FAIL: TestIntegration (31.66s)
>     --- FAIL: TestIntegration/golang-go (0.11s)
>         integration_test.go:243: initialize: jsonrpc2: code 0 message: unable to determine workspace's root Go import path: invalid canonical import path "go/constant" in file at path "/src/go/constant/value.go" (original rootPath is "git://github.com/golang/go?f75aafdf56dd90eab75cfeac8cf69358f73ba171") git://github.com/golang/go?f75aafdf56dd90eab75cfeac8cf69358f73ba171
> ```
> 
> Based on https://golang.org/doc/go1.4#canonicalimports, this doesn't look invalid to me https://sourcegraph.sgdev.org/github.com/golang/go@f75aafdf56dd90eab75cfeac8cf69358f73ba171/-/blob/src/go/constant/value.go?diff=f75aafdf56dd90eab75cfeac8cf69358f73ba171&utm_source=chrome-extension#L13
> 
> ```
> package constant // import "go/constant"
> ```
> 
> @-keegancsmith Weird, is this test failing on current master? This is likely a side-effect of not running all tests on CI. This is indeed a correct import path, but only for GOROOT. Is our GOROOT detection not working?
> 
> @-chrismwendt Yes, this is failing on master, but CI skips this test. I filed https://github.com/sourcegraph/sourcegraph/issues/1135 to track this.